<a href="https://cognitiveclass.ai/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkPY0101ENSkillsNetwork19487395-2021-01-01">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0101EN-SkillsNetwork/IDSNlogo.png" width="300" align="center">
</a>


<h1> HTTP and Requests</h1>

Estimated time needed: **30** minutes

## Objectives

After completing this lab you will be able to:

*   Understand HTTP
*   Handle HTTP Requests


<h2>Table of Contents</h2>

<div class="alert alert-block alert-info" style="margin-top: 20px">
    <ul>
        <li>
            <a href="#Overview-of-HTTP">Overview of HTTP </a>
            <ul>
                <li><a href="#Uniform-Resource-Locator:URL">Uniform Resource Locator:URL</a></li>
                 <li><a href="#Request">Request</a></li>
                <li><a href="#Response">Response</a></li>
            </ul>
        </li>
        <li>
            <a href="#Requests-in-Python">Requests in Python  </a>
            <ul>
                <li><a href="#Get-Request-with-URL-Parameters">Get Request with URL Parameters</a></li>
                <li><a href="#Post-Requests">Post Requests </a></li>

</ul>

</div>

<hr>


## Overview of HTTP


When you, the **client**, use a web page your browser sends an **HTTP** request to the **server** where the page is hosted. The server tries to find the desired **resource** by default "<code>index.html</code>". If your request is successful, the server will send the object to the client in an **HTTP response**. This includes information like the type of the **resource**, the length of the **resource**, and other information.

<p>
The figure below represents the process. The circle on the left represents the client, the circle on the right represents the Web server. The table under the Web server represents a list of resources stored in the web server. In  this case an <code>HTML</code> file, <code>png</code> image, and <code>txt</code> file .
</p>
<p>
The <b>HTTP</b> protocol allows you to send and receive information through the web including webpages, images, and other web resources. In this lab, we will provide an overview of the Requests library for interacting with the <code>HTTP</code> protocol.
</p


<div class="alert alert-block alert-info" style="margin-top: 20px">
         <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0101EN-SkillsNetwork/labs/Module%205/images/reqest_basics.png" width="750" align="center">

</div>


## Uniform Resource Locator:URL


Uniform resource locator (URL) is the most popular way to find resources on the web.  We can break the URL into three parts.

<ul>
    <li><b>Scheme</b>:- This is this protocol, for this lab it will always be <code>http://</code>  </li>
    <li><b> Internet address or  Base URL </b>:- This will be used to find the location here are some examples: <code>www.ibm.com</code> and  <code> www.gitlab.com </code> </li>
    <li><b>Route</b>:- Location on the web server for example: <code>/images/IDSNlogo.png</code> </li>
</ul>


You may also hear the term Uniform Resource Identifier (URI), URL are actually a subset of URIs. Another popular term is endpoint, this is the URL of an operation provided by a Web server.


## Request


The process can be broken into the <b>Request</b> and <b>Response </b> process.  The request using the get method is partially illustrated below. In the start line we have the <code>GET</code> method, this is an <code>HTTP</code> method. Also the location of the resource  <code>/index.html</code> and the <code>HTTP</code> version. The Request header passes additional information with an <code>HTTP</code> request:


<div class="alert alert-block alert-info" style="margin-top: 20px">
         <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0101EN-SkillsNetwork/labs/Module%205/images/reqest_messege.png" width="400" align="center">
</div>


When an <code>HTTP</code> request is made, an <code>HTTP</code> method is sent, this tells the server what action to perform.  A list of several <code>HTTP</code> methods is shown below. We will go over more examples later.


<div class="alert alert-block alert-info" style="margin-top: 20px">
         <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0101EN-SkillsNetwork/labs/Module%205/images/http_methods.png" width="400" align="center">
</div>


## Response


The figure below represents the response; the response start line contains the version number <code>HTTP/1.0</code>, a status code (200) meaning success, followed by a descriptive phrase (OK). The response header contains useful information. Finally, we have the response body containing the requested file, an <code> HTML </code> document.  It should be noted that some requests have headers.


<div class="alert alert-block alert-info" style="margin-top: 20px">
         <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0101EN-SkillsNetwork/labs/Module%205/images/response_message.png" width="400" align="center">
</div>


Some status code examples are shown in the table below, the prefix indicates the class. These are shown in yellow, with actual status codes shown in  white. Check out the following <a href="https://developer.mozilla.org/en-US/docs/Web/HTTP/Status?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkPY0101ENSkillsNetwork19487395-2021-01-01">link </a> for more descriptions.


<div class="alert alert-block alert-info" style="margin-top: 20px">
         <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0101EN-SkillsNetwork/labs/Module%205/images/status_code.png" width="300" align="center">
</div>


## Requests in Python


Requests is a Python Library that allows you to send <code>HTTP/1.1</code> requests easily. We can import the library as follows:


In [4]:
import requests

We will also use the following libraries:


In [5]:
import os
from PIL import Image
from IPython.display import IFrame

You can make a <code>GET</code> request via the method <code>get</code> to [www.ibm.com](http://www.ibm.com/?utm_source=Exinfluencer&utm_content=000026UJ&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkPY0101ENSkillsNetwork19487395-2021-01-01&utm_medium=Exinfluencer&utm_term=10006555):


In [6]:

url='https://www.ibm.com/'
r=requests.get(url)


We have the response object <code>r</code>, this has information about the request, like the status of the request. We can view the status code using the attribute <code>status_code</code>.


In [7]:
r.status_code

200

You can view the request headers:


In [8]:
print(r.request.headers)

{'User-Agent': 'python-requests/2.31.0', 'Accept-Encoding': 'gzip, deflate', 'Accept': '*/*', 'Connection': 'keep-alive', 'Cookie': '_abck=991D9874BAB247BA5D2DC6AF7D26EB79~-1~YAAQzunHF+2fQFSQAQAA/EHoYAyS8oi34A1ieLKd4r7jcFcHKEMg8SjzUySZoN15+y7PfMXEDUqBlkEGE0HU8/7R1bVl2NQc+u2wG+YNRKb4hupWZ5IO08/RpaowqN2cmJHgjmVNU8por5ChMXS56aVOhLrAOKdQXeQ8hGT5U9M79wcDTdqs+wJytawKy7itbNoWXZOoxo8qx4dLp6PGxrAqeWuS8UVbs2Gc+sRu6ZKlhhOtf+CW8PWQvnjhKK0tvg/b5YNASH4cPG6hQOuQgAhBSXAf3kkrbr0s2Mf8B69g+WV4JNhE14ZsP8m0ZXocDCTMSPPuCWYVuYdGVBbpsOAwc2EBO6VCIRxflXFRY1TR2VYXm7U=~-1~-1~-1; bm_sz=A00E9F850BFBA142EB0B99C9375EF8C1~YAAQzunHF+6fQFSQAQAA/EHoYBjuoNT2mwIxWVtMh2K0fcjbC/eKquRcMxOcbhnt7+ofK+wONJuSOqwHzK8GESHAv4BVBoiok0zM5vc4Z6vozIfnTW6hrOuMPFpm3eBbdXVZRqDFBCWm6DKJhN3In/yLiERpSmavtFlytaLUif8YaXnZAXpkKfH5rnlOHNoVR0q75fDg+TIMCi11X0pG8E6/5OudqMDYPcXJJdXizTWocuH7hUA850CK4Ndws9p9zQ44P0THPdifrRbsmW/48rwau2r0oI843EbH/w5dPtSln0cZ1VBKbg2rhWW3JQsIvmfLThuK6Mk1pYcgSJJnaY/ZeTTHY989kFM=~3355970~3159107'}


You can view the request body, in the following line, as there is no body for a get request we get a <code>None</code>:


In [9]:
print("request body:", r.request.body)

request body: None


You can view the <code>HTTP</code> response header using the attribute <code>headers</code>. This returns a python dictionary of <code>HTTP</code> response headers.


In [10]:
header=r.headers
print(r.headers)

{'Content-Security-Policy': 'upgrade-insecure-requests', 'x-frame-options': 'SAMEORIGIN', 'Last-Modified': 'Fri, 28 Jun 2024 22:05:07 GMT', 'ETag': 'W/"16990-61bfa70bf58d3-gzip"', 'Accept-Ranges': 'bytes', 'Content-Type': 'text/html;charset=utf-8', 'X-Content-Type-Options': 'nosniff', 'Cache-Control': 'max-age=600', 'Expires': 'Fri, 28 Jun 2024 22:22:32 GMT', 'X-Akamai-Transformed': '9 12611 0 pmb=mTOE,2', 'Content-Encoding': 'gzip', 'Date': 'Fri, 28 Jun 2024 22:12:32 GMT', 'Content-Length': '12805', 'Connection': 'keep-alive', 'Vary': 'Accept-Encoding', 'Strict-Transport-Security': 'max-age=31536000'}


We can obtain the date the request was sent using the key <code>Date</code>.


In [11]:
header['date']

'Fri, 28 Jun 2024 22:12:32 GMT'

<code>Content-Type</code> indicates the type of data:


In [12]:
header['Content-Type']

'text/html;charset=utf-8'

You can also check the <code>encoding</code>:


In [20]:
 r.encoding

As the <code>Content-Type</code> is <code>text/html</code> we can use the attribute <code>text</code> to display the <code>HTML</code> in the body. We can review the first 100 characters:


In [14]:
r.text[0:100]

'\n<!DOCTYPE HTML>\n<html lang="en-us">\n<head>\r\n    \r\n    \r\n    \r\n    \r\n    \r\n    \r\n    \r\n    <meta cha'

You can load other types of data for non-text requests, like images. Consider the URL of the following image:


In [15]:
# Use single quotation marks for defining string
url='https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0101EN-SkillsNetwork/IDSNlogo.png'

We can make a get request:


In [16]:
r=requests.get(url)

We can look at the response header:


In [17]:
print(r.headers)

{'Date': 'Fri, 28 Jun 2024 22:15:06 GMT', 'X-Clv-Request-Id': 'e72b0f35-78f2-463d-8ce3-4e02da9bec54', 'Server': 'Cleversafe', 'X-Clv-S3-Version': '2.5', 'Accept-Ranges': 'bytes', 'x-amz-request-id': 'e72b0f35-78f2-463d-8ce3-4e02da9bec54', 'ETag': '"8bb44578fff8fdcc3d2972be9ece0164"', 'Content-Type': 'image/png', 'Last-Modified': 'Wed, 16 Nov 2022 03:32:41 GMT', 'Content-Length': '78776'}


We can see the <code>'Content-Type'</code>


In [21]:
r.headers['Content-Type']

'image/png'

An image is a response object that contains the image as a <a href="https://docs.python.org/3/glossary.html?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkPY0101ENSkillsNetwork19487395-2021-01-01#term-bytes-like-object">bytes-like object</a>. As a result, we must save it using a file object. First, we specify the <u>file path and
name</u>


In [23]:
path=os.path.join(os.getcwd(),'image.png')

We save the file, in order to access the body of the response we use the attribute <code>content</code> then save it using the <code>open</code> function and write <code>method</code>:


In [24]:
with open(path,'wb') as f:
    f.write(r.content)

We can view the image:


In [ ]:
Image.open(path)

<h3>Question: Download a file </h3>


Consider the following URL.


<code>URL = <https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0101EN-SkillsNetwork/labs/Module%205/data/Example1.txt</code>


In [32]:


URL = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0101EN-SkillsNetwork/labs/Module%205/data/Example1.txt"
import requests
import os
response = requests.get(URL)
if r.status_code == 200:
    header=response.headers
    print(header)
    print(response.headers['Content-Type'])
    # print(response.text[0:100])
    if response.headers['Content-Type'] == 'text/plain':
        path=os.path.join(os.getcwd(),'example1.txt')
        print(path)
        print(os.getcwd())
        with open(path,'w') as f:
            f.write(response.text)
else:
    print("Error")





{'Date': 'Fri, 28 Jun 2024 22:49:54 GMT', 'X-Clv-Request-Id': 'ea15446c-52bb-4d5d-b706-61e7433729ae', 'Server': 'Cleversafe', 'X-Clv-S3-Version': '2.5', 'Accept-Ranges': 'bytes', 'x-amz-request-id': 'ea15446c-52bb-4d5d-b706-61e7433729ae', 'Cache-Control': 'max-age=0,public', 'ETag': '"522bc63a850f3569b8924c51131d5fe4"', 'Content-Type': 'text/plain', 'Last-Modified': 'Wed, 12 Oct 2022 22:36:17 GMT', 'Content-Length': '45'}
text/plain
/content/example1.txt
/content


Write the commands to download the txt file in the given link.


In [48]:


URL = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0101EN-SkillsNetwork/labs/Module%205/data/Example1.txt"
import requests
import os
import re
response = requests.get(URL)
if r.status_code == 200:
    header=response.headers
    print(header)
    print(response.headers['Content-Type'])
    # print(response.text[0:100])
    if response.headers['Content-Type'] == 'text/plain':
        path1 = r"C:\Users\ranas\OneDrive\Documents\Npower\Python\Code\tmp"
        path = os.path.join(path1, 'example1.txt')
        # path1 = r"C:\Users\ranas\OneDrive\Documents\Npower\Python\Code\tmp"
        # path = path1 + r"\example1.txt"
        with open(path,'wb') as f:
            f.write(response.text)
else:
    print("Error")

url='https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0101EN-SkillsNetwork/labs/Module%205/data/Example1.txt'
path=os.path.join(os.getcwd(),'example1.txt')
r=requests.get(url)
print(path)
print(os.getcwd())
with open(path,'wb') as f:
    f.write(r.content)





{'Date': 'Fri, 28 Jun 2024 23:27:00 GMT', 'X-Clv-Request-Id': '864e9e8a-89bf-405f-85e8-c73210a32297', 'Server': 'Cleversafe', 'X-Clv-S3-Version': '2.5', 'Accept-Ranges': 'bytes', 'x-amz-request-id': '864e9e8a-89bf-405f-85e8-c73210a32297', 'Cache-Control': 'max-age=0,public', 'ETag': '"522bc63a850f3569b8924c51131d5fe4"', 'Content-Type': 'text/plain', 'Last-Modified': 'Wed, 12 Oct 2022 22:36:17 GMT', 'Content-Length': '45'}
text/plain


FileNotFoundError: [Errno 2] No such file or directory: 'C:\\Users\\ranas\\OneDrive\\Documents\\Npower\\Python\\Code\\tmp/example1.txt'

<details><summary>Click here for the solution</summary>

```python
url='https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0101EN-SkillsNetwork/labs/Module%205/data/Example1.txt'
path=os.path.join(os.getcwd(),'example1.txt')
r=requests.get(url)
with open(path,'wb') as f:
    f.write(r.content)

```

</details>


## Get Request with URL Parameters


You can use the <b>GET</b> method to modify the results of your query, for example retrieving data from an API. We send a <b>GET</b> request to the  server. Like before we have the <b>Base URL</b>, in the <b>Route</b> we append <code>/get</code>, this indicates we would like to preform a <code>GET</code> request.


The Base URL is for <code>[http://httpbin.org/](http://httpbin.org/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkPY0101ENSkillsNetwork19487395-2021-01-01)</code> is a simple HTTP Request & Response Service. The <code>URL</code> in Python is given by:


In [59]:
url_get='http://httpbin.org/get'

A <a href="https://en.wikipedia.org/wiki/Query_string?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkPY0101ENSkillsNetwork19487395-2021-01-01">query string</a> is a part of a uniform resource locator (URL), this sends other information to the web server. The start of the query is a <code>?</code>, followed by a series of parameter and value pairs, as shown in the table below. The first parameter name is <code>name</code> and the value is <code>Joseph</code>. The second parameter name is <code>ID</code> and the Value is <code>123</code>. Each pair, parameter, and value is separated by an equals sign, <code>=</code>.
The series of pairs is separated by the ampersand <code>&</code>.


<div class="alert alert-block alert-info" style="margin-top: 20px">
         <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0101EN-SkillsNetwork/labs/Module%205/images/query_string.png" width="500" align="center">
</div>


To create a Query string, add a dictionary. The keys are the parameter names and the values are the value of the Query string.


In [61]:
payload={"name":"Joseph","ID":"123"}

Then passing the dictionary <code>payload</code> to the <code>params</code> parameter of the <code> get()</code> function:


In [62]:
r=requests.get(url_get,params=payload)

We can print out the <code>URL</code> and see the name and values.


In [68]:
r.url

'http://httpbin.org/get?name=Joseph&ID=123'

There is no request body.


In [64]:
print("request body:", r.request.body)

request body: None


We can print out the status code.


In [65]:
print(r.status_code)

200


We can view the response as text:


In [66]:
print(r.text)

{
  "args": {
    "ID": "123", 
    "name": "Joseph"
  }, 
  "headers": {
    "Accept": "*/*", 
    "Accept-Encoding": "gzip, deflate", 
    "Host": "httpbin.org", 
    "User-Agent": "python-requests/2.31.0", 
    "X-Amzn-Trace-Id": "Root=1-667f487a-2646ec3e77872cec64470076"
  }, 
  "origin": "34.125.198.97", 
  "url": "http://httpbin.org/get?name=Joseph&ID=123"
}



We can look at the <code>'Content-Type'</code>.


In [67]:
r.headers['Content-Type']

'application/json'

As the content <code>'Content-Type'</code> is in the <code>JSON</code> format we can use the method <code>json()</code>, it returns a Python <code>dict</code>:


In [57]:
r.json()

{'args': {'ID': '123', 'name': 'Joseph'},
 'headers': {'Accept': '*/*',
  'Accept-Encoding': 'gzip, deflate',
  'Host': 'httpbin.org',
  'User-Agent': 'python-requests/2.31.0',
  'X-Amzn-Trace-Id': 'Root=1-667f4702-133281e00794cde338a13164'},
 'origin': '34.125.198.97',
 'url': 'http://httpbin.org/get?name=Joseph&ID=123'}

The key <code>args</code> has the name and values:


In [58]:
r.json()['args']

{'ID': '123', 'name': 'Joseph'}

## Post Requests


Like a <code>GET</code> request, a <code>POST</code> is used to send data to a server, but the <code>POST</code> request sends the data in a request body. In order to send the Post Request in Python, in the <code>URL</code> we change the route to <code>POST</code>:


In [69]:
url_post='http://httpbin.org/post'

This endpoint will expect data as a file or as a form. A form is convenient way to configure an HTTP request to send data to a server.


To make a <code>POST</code> request we use the <code>post()</code> function, the variable <code>payload</code> is passed to the parameter <code> data </code>:


In [70]:
r_post=requests.post(url_post,data=payload)

Comparing the URL from the response object of the <code>GET</code> and <code>POST</code> request we see the <code>POST</code> request has no name or value pairs.


In [71]:
print("POST request URL:",r_post.url )
print("GET request URL:",r.url)

POST request URL: http://httpbin.org/post
GET request URL: http://httpbin.org/get?name=Joseph&ID=123


We can compare the <code>POST</code> and <code>GET</code> request body, we see only the <code>POST</code> request has a body:


In [72]:
print("POST request body:",r_post.request.body)
print("GET request body:",r.request.body)

POST request body: name=Joseph&ID=123
GET request body: None


We can view the form as well:


In [73]:
r_post.json()['form']

{'ID': '123', 'name': 'Joseph'}

There is a lot more you can do. Check out <a href="https://requests.readthedocs.io/en/master/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkPY0101ENSkillsNetwork19487395-2021-01-01">Requests </a> for more.


<hr>


## Authors

<p><a href="https://www.linkedin.com/in/joseph-s-50398b136/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkPY0101ENSkillsNetwork19487395-2021-01-01" target="_blank">Joseph Santarcangelo</a> <br>A Data Scientist at IBM, and holds a PhD in Electrical Engineering. His research focused on using Machine Learning, Signal Processing, and Computer Vision to determine how videos impact human cognition. Joseph has been working for IBM since he completed his PhD.</p>


### Other Contributors

<a href="https://www.linkedin.com/in/jiahui-mavis-zhou-a4537814a?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkPY0101ENSkillsNetwork19487395-2021-01-01">Mavis Zhou</a>


## Change Log

| Date (YYYY-MM-DD) | Version | Changed By | Change Description           |
| ----------------- | ------- | ---------- | ---------------------------- |
| 2023-11-02 | 2.4 | Abhishek Gagneja | Updated instructions |
| 2023-06-07        | 2.3     |Akansha Yadav| Spell Check                 |
| 2021-12-20        | 2.1     | Malika     | Updated the links            |
| 2020-09-02        | 2.0     | Simran     | Template updates to the file |
|                   |         |            |                              |
|                   |         |            |                              |

## <h3 align="center"> © IBM Corporation 2023. All rights reserved. <h3/>
